### PR TITLE
add a naive NestedLoopsJoin: executable operator, physical operator, …

### PR DESCRIPTION
**Pull request by chengsijin0817**
…and unit test (issue #14)

### CODE REVIEW

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+		final GenericIntermediateResultBlockImpl input2 = new GenericIntermediateResultBlockImpl();
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x1,
+				var2, y1,
+				var3, z1) );
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var1, x2,
+				var2, y2,
+				var3, z2) );
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java:135]: Please add another unit test where the join is on two variables and input solution mappings have the same value for one of the variables but not for the other (i.e., there are no join partners).

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		input2.add( SolutionMappingUtils.createSolutionMapping(
+				var2, y1,
+				var3, z1) );
+
+		final Iterator<SolutionMapping> it = runTest(input1, input2);
+
+		assertFalse( it.hasNext() );
+	}
+
+	protected void _joinWithEmptyInput2() {
```
Comment (hartig) [inline src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java:240]: Please add two more unit tests where one of the two inputs has one solution mapping which happens to be the empty solution mapping (which is compatible with any solution mapping).

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        return 1;
+    }
+
+    @Override
+    public boolean requiresCompleteChild1InputFirst() {
+        return true;
+    }
+
+    @Override
+    public void processBlockFromChild1(IntermediateResultBlock input, IntermediateResultElementSink sink, ExecutionContext execCxt) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java:25]:
```suggestion
    public void processBlockFromChild1( final IntermediateResultBlock input, final IntermediateResultElementSink sink, final ExecutionContext execCxt ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+            inputLHS.add(sm);
+        }
+    }
+
+    @Override
+    public void wrapUpForChild1(IntermediateResultElementSink sink, ExecutionContext execCxt) {
+    }
+
+    @Override
+    public void processBlockFromChild2(IntermediateResultBlock input, IntermediateResultElementSink sink, ExecutionContext execCxt) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java:36]:
```suggestion
    public void processBlockFromChild2( final IntermediateResultBlock input, final IntermediateResultElementSink sink, final ExecutionContext execCxt ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            for ( final SolutionMapping smL : inputLHS ) {
+                if ( SolutionMappingUtils.compatible(smL,smR) ) {
+                    sink.send( SolutionMappingUtils.merge(smL,smR) );
+                }
+            }
+        }
+    }
+
+    @Override
+    public void wrapUpForChild2(IntermediateResultElementSink sink, ExecutionContext execCxt) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpNaiveNestedLoopsJoin.java:47]:
```suggestion
    public void wrapUpForChild2( final IntermediateResultElementSink sink, final ExecutionContext execCxt ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+	protected final BinaryLogicalOp lop;
+
+	protected BasePhysicalOpBinaryInputJoin(BinaryLogicalOp lop) {
+		assert lop != null;
+		assert (lop instanceof LogicalOpJoin);
+		this.lop = lop;
+	}
+
+	@Override
+	public ExpectedVariables getExpectedVariables(ExpectedVariables... inputVars) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java:23]:
```suggestion
	public ExpectedVariables getExpectedVariables( final ExpectedVariables... inputVars ) {
```

**Review by hartig — CHANGES_REQUESTED** [review]

**Comment by hartig** [comment]
I have reviewed this PR. See my detailed comments above. Please address them. Many of them can be addressed in the Web interface of github by simply accepting my suggested changes in the Web interface.

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java (hunk 2 above)
Comment (chengsijin0817) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java:240]: I added two more unit tests in TestsForJoinAlgorithms, please see commit f43e5b8

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+import se.liu.ida.hefquin.engine.queryplan.logical.BinaryLogicalOp;
+import se.liu.ida.hefquin.engine.queryplan.logical.impl.LogicalOpJoin;
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOpForLogicalOp;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class BasePhysicalOpBinaryInputJoin implements BinaryPhysicalOpForLogicalOp
+{
+	protected final BinaryLogicalOp lop;
```
[the remark this answers is not in the document]
Comment (chengsijin0817) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java]: The type here is updated to LogicalOpJoin. (commit f43e5b8)

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+import se.liu.ida.hefquin.engine.queryplan.physical.BinaryPhysicalOpForLogicalOp;
+
+import java.util.HashSet;
+import java.util.Set;
+
+public abstract class BasePhysicalOpBinaryInputJoin implements BinaryPhysicalOpForLogicalOp
+{
+	protected final LogicalOpJoin lop;
+
+	protected BasePhysicalOpBinaryInputJoin(LogicalOpJoin lop) {
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java:16]:
```suggestion
	protected BasePhysicalOpBinaryInputJoin( final LogicalOpJoin lop ) {
```

File: src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/LogicalToPhysicalPlanConverterImpl.java [hunk 9]
```diff
@@ -183,7 +179,7 @@ protected UnaryPhysicalOp convertBGPAdd( final LogicalOpBGPAdd lop ) {
 
 	protected BinaryPhysicalOp convertJoin( final LogicalOpJoin lop ) {
 		// TODO implement convertJoin (for which we need some physical operator for join)
```
Comment (hartig) [inline src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/LogicalToPhysicalPlanConverterImpl.java:181]:
```suggestion
```

File: src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (hartig) [reply src/main/java/se/liu/ida/hefquin/engine/queryplan/physical/impl/BasePhysicalOpBinaryInputJoin.java]: Good. Thanks.

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java (hunk 2 above)
Comment (hartig) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java:240]: Thanks for these two additional unit tests!

File: src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java (hunk 1 above)
Comment (hartig) [reply src/test/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/TestsForJoinAlgorithms.java:134]: Thanks for adding that other unit test!

**Review by hartig — APPROVED** [review]

**Comment by hartig** [comment]
Thanks for addressing my comments. I will merge this PR now.
